src/lib.rs

```diff
@@ -151,58 +156,42 @@ pub fn generate(mut args: Args) -> Result<()> {
-            let template_folder = resolve_template_dir(&template_base_dir, args)?;
-            (template_base_dir, template_folder, branch)
-        }
-        (None, Some(_)) => {
-            let template_base_dir = copy_path_template_into_temp(args)?;
-            let branch = args.branch.clone().unwrap_or_else(|| String::from("main"));
-            let template_folder =
-                auto_locate_template_dir(template_base_dir.path(), prompt_for_variable)?;
-            (template_base_dir, template_folder, branch)
-        }
-        _ => bail!(
-            "{} {} {} {} {}",
-            emoji::ERROR,
-            style("Please specify either").bold(),
-            style("--git <repo>").bold().yellow(),
-            style("or").bold(),
-            style("--path <path>").bold().yellow(),
-        ),
-    };
+fn prepare_local_template(
+    source_template: &UserParsedInput,
+) -> Result<(TempDir, PathBuf, String), anyhow::Error> {
+    let (temp_dir, branch) = get_source_template_into_temp(source_template.location())?;
+    let template_folder = resolve_template_dir(&temp_dir, source_template.subfolder())?;
 
-    Ok((template_base_dir, template_folder, branch))
+    Ok((temp_dir, template_folder, branch))
 }
 
-fn check_cargo_generate_version(template_config: &Config) -> Result<(), anyhow::Error> {
-    if let Config {
-        template:
-            Some(config::TemplateConfig {
-                cargo_generate_version: Some(requirement),
-                ..
-            }),
-        ..
-    } = template_config
-    {
-        let version = semver::Version::parse(env!("CARGO_PKG_VERSION"))?;
-        if !requirement.matches(&version) {
-            bail!(
-                "{} {} {} {} {}",
-                emoji::ERROR,
-                style("Required cargo-generate version not met. Required:")
-                    .bold()
-                    .red(),
-                style(requirement).yellow(),
-                style(" was:").bold().red(),
-                style(version).yellow(),
-            );
+fn get_source_template_into_temp(
+    template_location: &TemplateLocation,
+) -> Result<(TempDir, String)> {
+    let temp_dir: TempDir;
+    let branch: String;
+    match template_location {
+        TemplateLocation::Git(git) => {
+            let clone_fn = if std::env::var("CARGO_GENERATE_GIT_WITH_CLI")
+                .map(|v| v.to_lowercase() == "true")
+                .unwrap_or(false)
+            {
+                git::clone_git_using_cmd
+            } else {
+                git::clone_git_template_into_temp
+            };
+
+            let (temp_dir2, branch2) = clone_fn(git.url(), git.branch(), git.identity())?;
+            temp_dir = temp_dir2;
+            branch = branch2;
         }
-    }
-    Ok(())
+        TemplateLocation::Path(path) => {
+            temp_dir = copy_path_template_into_temp(path)?;
+            branch = String::from(DEFAULT_BRANCH); // FIXME is here any reason to set branch when path is used?
```

Comment:
When `--path` is used we copy template in tmp and returning `branch`. But there is no branch actually

Reply:
the normal git clone mechanism yields the branch, so this must also provide a branch

Reply:
Oh. I mean this should probably return `Option<String>` instead of `String`. If the branch is returned it's surprise that there is actually no git repository. 

And it would be more clear what's going on [here]( https://github.com/xoac/cargo-generate/blob/7e19fd8673c43fb24b19e96e876b0b621c427644/src/lib.rs#L146):
```rust
    if !args.init || args.force_git_init {
        info!("{}", style("Initializing a fresh Git repository").bold());
        args.vcs
            .initialize(&project_dir, branch, args.force_git_init)?;
    }
```
and replace `branch` with `branch.unwrap_or(DEFAULT_BRANCH)` 

----

For me the whole behavior here is surprise. I would expect that if I call `cargo gen --branch web3 example-org/example-template` it will init `main` branch instead of `web3`. What you think about `--init--branch` flag with default value `main` for initialize git repository? (I mean in other PR).